jgrapht-core/src/main/java/org/jgrapht/alg/scoring/Coreness.java

```diff
@@ -0,0 +1,170 @@
+ * least k. Equivalently, it is one of the connected components of the subgraph of G formed by
+ * repeatedly deleting all vertices of degree less than k. A vertex u has coreness c if it belongs
+ * to a c-core but not to any (c+1)-core.
+ *
+ * <p>
+ * If a non-empty k-core exists, then, clearly, G has
+ * <a href="https://en.wikipedia.org/wiki/Degeneracy_(graph_theory)">degeneracy</a> at least k, and
+ * the degeneracy of G is the largest k for which G has a k-core.
+ *
+ * <p>
+ * As described in the following paper
+ * <ul>
+ * <li>D. W. Matula and L. L. Beck. Smallest-last ordering and clustering and graph coloring
+ * algorithms. Journal of the ACM, 30(3):417--427, 1983.</li>
+ * </ul>
+ * it is possible to find a vertex ordering of a finite graph G that optimizes the coloring number
+ * of the ordering, in linear time, by using a bucket queue to repeatedly find and remove the vertex
+ * of smallest degree.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Dimitrios Michail
+ */
+public final class Coreness<V, E>
+    implements VertexScoringAlgorithm<V, Integer>
+{
+    private final UndirectedGraph<V, E> g;
+    private Map<V, Integer> scores;
+    private int degeneracy;
+
+    /**
+     * Constructor
+     * 
+     * @param g the input graph
+     */
+    public Coreness(UndirectedGraph<V, E> g)
+    {
+        this.g = Objects.requireNonNull(g, "Graph cannot be null");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Map<V, Integer> getScores()
+    {
+        lazyRun();
+        return Collections.unmodifiableMap(scores);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Integer getVertexScore(V v)
+    {
+        if (!g.containsVertex(v)) {
+            throw new IllegalArgumentException("Cannot return score of unknown vertex");
+        }
+        lazyRun();
+        return scores.get(v);
+    }
+
+    /**
+     * Compute the degeneracy of a graph.
+     * 
+     * <p>
+     * The degeneracy of a graph is the smallest value of k for which it is k-degenerate. In graph
+     * theory, a k-degenerate graph is an undirected graph in which every subgraph has a vertex of
+     * degree at most k: that is, some vertex in the subgraph touches k or fewer of the subgraph's
+     * edges.
+     * 
+     * @return the degeneracy of a graph
+     */
+    public int getDegeneracy()
+    {
+        lazyRun();
+        return degeneracy;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void lazyRun()
+    {
+        if (scores != null) {
+            return;
+        }
+
+        scores = new HashMap<>();
+        degeneracy = 0;
+
+        /*
+         * Initialize buckets
+         */
+        int n = g.vertexSet().size();
+        Set<V>[] buckets = (Set<V>[]) Array.newInstance(Set.class, n + 1);
+        for (int i = 0; i <= n; i++) {
+            buckets[i] = new HashSet<>();
+        }
+
+        int minDegree = n + 1;
```

Comment:
minDegree = n + 1;

This raises some questions:
- do you allow graphs with self loops? If so, then the max degree of any vertex in a complete graph + self loop = n
- if you do not allow self loops, then the max degree of any vertex = n-1
- what about pseudo graphs where the degree of a vertex can be arbitrarily large? If this class doesn't support pseudo graphs, then we should probably make a remark about this in the class description?